core/src/data_sources/data_source.rs

```diff
@@ -605,23 +607,112 @@ impl DataSource {
 
         let now = utils::now();
         let mut offset: usize = 0;
+        let current_doc = store
+            .load_data_source_document(
+                &self.project,
+                &self.data_source_id(),
+                &document_id.to_string(),
+                &None,
+            )
+            .await?;
+
+        let mut should_embbed = true;
+        match current_doc {
+            Some(current_doc) => {
+                if current_doc.hash == document_hash {
+                    should_embbed = false;
+                }
+            }
+            _ => {
+                should_embbed = true;
+            }
+        };
+
         // Embed batched chunks sequentially.
         for chunk in chunked_splits {
-            let r = EmbedderRequest::new(
-                self.config.provider_id.clone(),
-                &self.config.model_id,
-                chunk.iter().map(|s| s.as_str()).collect::<Vec<_>>(),
-                self.config.extras.clone(),
-            );
-
-            let v = match r.execute(credentials.clone()).await {
-                Ok(v) => v,
-                Err(e) => Err(anyhow!("DataSource chunk embedding error: {}", e))?,
-            };
+            println!("Should embbed {:?}", should_embbed);
+            if should_embbed == false {
+                for s in chunk.iter() {
```

Comment:
`chunked_splits` is used to batch call the embedder, but you can surely retrieve all the chunks from the document directly from qdrant. the fork should be higher up.